cli/tools/registry/pm/outdated/interactive.rs

```diff
@@ -0,0 +1,342 @@
+  packages: Vec<PackageInfo>,
+  currently_selected: usize,
+  checked: HashSet<usize>,
+
+  name_width: usize,
+  current_width: usize,
+  start_row: u16,
+}
+
+impl State {
+  fn new(packages: Vec<PackageInfo>) -> anyhow::Result<Self> {
+    let name_width = packages
+      .iter()
+      .map(|p| p.name.len() + p.kind.scheme().len() + 1)
+      .max()
+      .unwrap_or_default();
+    let current_width = packages
+      .iter()
+      .map(|p| p.current_version.len())
+      .max()
+      .unwrap_or_default();
+
+    let mut packages = packages;
+    packages
+      .sort_by(|a, b| a.location.file_path().cmp(&b.location.file_path()));
+
+    Ok(Self {
+      packages,
+      currently_selected: 0,
+      checked: HashSet::new(),
+
+      name_width,
+      current_width,
+      start_row: cursor::position()?.1,
+    })
+  }
+
+  fn render<W: std::io::Write>(&self, out: &mut W) -> anyhow::Result<()> {
+    use cursor::MoveTo;
+    use style::Print;
+    use style::PrintStyledContent;
+
+    crossterm::queue!(
+      out,
+      MoveTo(0, self.start_row),
+      terminal::Clear(terminal::ClearType::FromCursorDown),
+      PrintStyledContent("?".blue()),
+      Print(" Select which packages to update (<space> to select, ↑/↓/j/k to navigate, enter to accept, <Ctrl-c> to cancel)")
+    )?;
+
+    let base = self.start_row + 1;
+
+    for (i, package) in self.packages.iter().enumerate() {
+      if self.currently_selected == i {
+        crossterm::queue!(
+          out,
+          MoveTo(0, base + (self.currently_selected as u16)),
+          PrintStyledContent("❯".blue()),
+          Print(' '),
+        )?;
+      }
+      let checked = self.checked.contains(&i);
+      let selector = if checked { "●" } else { "○" };
+      crossterm::queue!(
+        out,
+        MoveTo(2, base + (i as u16)),
+        Print(selector),
+        Print(" "),
+      )?;
+
+      if self.currently_selected == i {
+        out.queue(style::SetStyle(
+          style::ContentStyle::new().on_black().white().bold(),
+        ))?;
+      }
+      let want = &package.new_version;
+      let new_version_highlight =
+        highlight_new_version(&package.current_version, want)?;
+      // let style = style::PrintStyledContent()
+      crossterm::queue!(
+        out,
+        Print(format!(
+          "{:<name_width$} {:<current_width$} -> {}",
+          format!(
+            "{}{}{}",
+            deno_terminal::colors::gray(package.kind.scheme()),
+            deno_terminal::colors::gray(":"),
+            package.name
+          ),
+          package.current_version,
+          new_version_highlight,
+          name_width = self.name_width + 2,
+          current_width = self.current_width
+        )),
+      )?;
+      // out.queue(Print(&package.package.name))?;
+      if self.currently_selected == i {
+        out.queue(style::ResetColor)?;
+      }
+    }
+
+    out.queue(MoveTo(0, base + self.packages.len() as u16))?;
+
+    out.flush()?;
+
+    Ok(())
+  }
```

Comment:
It might be better to switch this to https://github.com/dsherret/console_static_text with [deno_terminal](https://crates.io/crates/deno_terminal) because then we don't need the crossterm dependency. Additionally, this is already used by the draw thread for progress bars: https://github.com/denoland/deno/blob/main/cli/util/draw_thread.rs and in the future we could hook it up so that you can do stuff like show progress bars at the same time as having interaction https://david.deno.dev/posts/console-static-text/

That said, we probably still want something for handling key presses though maybe the repl has some code that does that?

Reply:
The repl uses `rustyline` which has its own keypress handling AFAICT.

Did a quick check, this PR adds about 66 KiB to the binary - though not sure how much of that is crossterm. This does also add an extra version of `mio` to our dependency tree (though that's because we're on an old tokio version – when we upgrade tokio we'll be able to share it with crossterm)

Reply:
For unix it probably wouldn't be too hard to roll our own, but no idea about windows

Reply:
I think at least here we should use deno_terminal and console_static_text to prevent the code we don't use ending up in the executable. Maybe something for a follow-up though.

Reply:
For deno_terminal, it would help us re-use the logic we have for NO_COLOR and TERM styling, but I think crossterm handles that as well: https://github.com/denoland/deno_terminal/blob/main/src/colors.rs